coding/python_testing.rst

```diff
@@ -4,6 +4,156 @@ Python Unit Testing
+The file is reproduced here:
+
+.. literalinclude:: unit_test_snippets/testExecutables.py
+   :linenos:
+   :language: python
+   :emphasize-lines: 8-9
+
+The :data:`EXECUTABLES` variable can be a tuple containing the names of the executables to be run (relative to the directory containing the test file).
+``None`` indicates that the test script should discover the executables in the same directory as that containing the test file.
+The call to :meth:`~lsst.utils.tests.ExecutablesTestCase.create_executable_tests` initiates executable discovery and creates a test for each executable that is found.
+
+In some cases an explicit test has to be written either because some precondition has to be met before the test will stand a chance of running or because some arguments have to be passed to the executable.
+To support this the :meth:`~lsst.utils.tests.ExecutableTestCase.assertExecutable` method is available:
+
+.. code-block:: python
+
+   def testBinary(self):
+       self.assertExecutable("binary1", args=None,
+                             root_dir=os.path.dirname(__file__))
+
+Where ``binary1`` is the name of the executable relative to the root directory specified in the ``root_dir`` optional argument.
+Arguments can be provided to the ``args`` keyword parameter in the form of a sequence of arguments in a list or tuple.
+
+.. note::
+   The LSST codebase is currently in transition such that :mod:`~lsst.sconsUtils` will run executables
+   itself as well as running Python test scripts that run executables.
+   Do not worry about this duplication of test running.
+   When the codebase has migrated to consistently use the testing scheme described in this section :mod:`~lsst.sconsUtils` will be modified to disable the duplicate testing.
+
+
+Memory and file descriptor leak testing
+---------------------------------------
+
+:class:`lsst.utils.tests.MemoryTestCase` is used to detect memory leaks in C++ objects and leaks in file descriptors.
+:class:`~lsst.utils.tests.MemoryTestCase` should be used in *all* test files, even if C++ code is not explicitly referenced.
+
+This example shows the basic structure of an LSST Python unit test module,
+including :class:`~lsst.utils.tests.MemoryTestCase`:
+
+.. literalinclude:: unit_test_snippets/unittest_runner_example.py
+   :linenos:
+   :language: python
+   :emphasize-lines: 6, 16, 21
+
+Note that :class:`~lsst.utils.tests.MemoryTestCase` must always be the
+final test suite.
+For the memory test to function properly the :func:`lsst.utils.tests.init()` function must be invoked before any of the tests in the class are executed.
+Since LSST test scripts are required to run properly from the command-line and when called from within `pytest`_, the :func:`~lsst.utils.tests.init()` function has to be called twice: once in the :func:`setup_module` function that is called by `pytest`_ whenever a test module is loaded, and also just before the call to :func:`unittest.main()` call.
```

Comment:
I find this is confusing. I think you are just warning us that it will automatically be called twice py pytest if we follow your rules. Surely it only has to appear once in each test file? If so, please clarify the text; perhaps turn it into a parenthetical remark.